src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c

```diff
@@ -0,0 +1,626 @@
+          response message in the previous step.
+    - In the successful case, the service responds on the $dps/registrations/res/200/?$rid={request_id} topic.
+    - The payload of the response contains the RegistrationOperationStatus object.
+    - The device should keep polling the service if the response code is 202 after a delay equal to the retry-after period.
+    - The device registration operation is successful if the service returns a 200 status code.
+*/
+
+bool DeviceRegistration_DoWork(/*const char* request_id, const char* connect_json*/)
+{
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_REGISTERED)
+    {
+        return true;
+    }
+
+    time_t nowTime = ADUC_GetTimeSinceEpochInSeconds();
+
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_REGISTERING)
+    {
+        if (nowTime > s_moduleState.nextRegisterAttemptTime)
+        {
+            SetRegisterState(ADPS_REGISTER_STATE_UNKNOWN, "time out");
+        }
+    }
+
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_WAIT_TO_POLL)
+    {
+        if (nowTime > s_moduleState.nextRegisterAttemptTime)
+        {
+            char* pollTopic = ADUC_StringFormat(
+                "$dps/registrations/GET/iotdps-get-operationstatus/?$rid=%s&operationId=%s",
+                s_moduleState.settings.registrationId,
+                s_moduleState.operationId);
+            if (pollTopic == NULL)
+            {
+                Log_Error("Cannot allocate memory for registration status polling topic");
+            }
+            else
+            {
+                int mid = 0;
+                int mqtt_ret = ADUC_CommunicationChannel_MQTT_Publish(
+                    pollTopic,
+                    &mid,
+                    0,
+                    NULL,
+                    s_moduleState.settings.mqttSettings.qos,
+                    false /* retain */,
+                    NULL /* props */);
+
+                if (mqtt_ret != MOSQ_ERR_SUCCESS)
+                {
+                    Log_Error("Failed to publish registration polling request (%d)", mqtt_ret);
+                    // TODO (nox-msft) : determine whether to retry or bail
+                }
+                else
+                {
+                    SetRegisterState(ADPS_REGISTER_STATE_POLLING, "publishing message");
+                    s_moduleState.nextRegisterAttemptTime = nowTime + ADPS_DEFAULT_REGISTER_REQUEST_DELAY_SECONDS;
+                    s_moduleState.registerRetries++;
+                }
+            }
+            free(pollTopic);
+        }
+        else
+        {
+            goto done;
+        }
+    }
+
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_FAILED)
+    {
+        // TODO (nox-msft) : determine whether to retry or bail
+        return false;
+    }
+
+    if (s_moduleState.registerState == ADPS_REGISTER_STATE_UNKNOWN && nowTime > s_moduleState.nextRegisterAttemptTime)
+    {
+        char topic_name[256];
```

Comment:
Magic number 256.  NOTE: Mqtt topic length is represented by 2 bytes, so can be up to 2^16 -1 = 65535 length, so it should be allocated on the heap as that's a bit large for the the stack.  Is 256 a limitation of mosquitto client, or just arbitrarily chosen?  http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718039